### PR TITLE
GRO-1585: validation for expectsFutureIncomeDecrease and reason

### PR DESCRIPTION
**Pull request by WeiLu1**
#### Description
[GRO-1585](https://selina.atlassian.net/browse/GRO-1585)
Add conditional validation for expectsFutureIncomeDecreaseReason.
If expectsFutureIncomeDecrease is true then the expectsFutureIncomeDecreaseReason must be one of the following enum values provided.

#### Background Context
Integrate with Ocean Finance.

#### Additional Information

[GRO-1585]: https://selina.atlassian.net/browse/GRO-1585?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by selina-bot-user** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **weilu1**
You can assign the PR to them by writing `/assign @weilu1` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=Selina-Finance%2Fms-lending-api).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/Selina-Finance/ms-lending-api/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by selina-bot-user** [comment]
@WeiLu1: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pr | dd6c0107b68d6603fcce9d959502165ed698c60f | [link](http://dashboard-jx.selinadev.co.uk/namespaces/jx/pipelineruns/finance-ms-lending-api-pr-163-pr-54q2h) | `/test pr`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by selina-bot-user** [comment]
:star: PR built and available in a preview **selina-finance-ms-lending-api-pr-163** [here](http://ms-lending-api-pr163.selinadev.co.uk)

**Comment by selina-bot-user** [comment]
:star: PR built and available in a preview **selina-finance-ms-lending-api-pr-163** [here](http://ms-lending-api-pr163.selinadev.co.uk)

File: src/main/java/com/selina/lending/internal/dto/IncomeDto.java [hunk 1]
```diff
@@ -21,17 +21,43 @@
...
 @Builder
-@Value
+@Data
+@Conditional(selected = "expectsFutureIncomeDecrease", values = {"true"}, required = {"expectsFutureIncomeDecreaseReason"})
 public class IncomeDto {
     @Valid
     List<IncomeItemDto> income;
     Boolean doesNotHaveAnyIncome;
     Boolean expectsFutureIncomeDecrease;
+    @Schema(implementation = ExpectsFutureIncomeDecreaseReasons.class, description = "the reason if  expectsFutureIncomeDecrease is true")
```
[the remark this answers is not in the document]
Comment (WeiLu1) [reply src/main/java/com/selina/lending/internal/dto/IncomeDto.java:38]: my bad, will fix

[the remark this answers is not in the document]
Comment (WeiLu1) [reply src/main/java/com/selina/lending/internal/dto/IncomeDto.java]: done

File: src/test/java/com/selina/lending/api/controller/DIPControllerValidationTest.java [hunk 2]
```diff
@@ -644,13 +644,42 @@ void shouldGiveValidationErrorWhenCreateDipApplicationAndSpecifyingIncorrectInco
...
-                    .contentType(APPLICATION_JSON))
-                 // Then
+                        .contentType(APPLICATION_JSON))
+                // Then
+                .andExpect(jsonPath("$.violations[0].field").value("applicants[0].income.income[0].frequency"))
+                .andExpect(jsonPath("$.violations[0].message").value("value is not valid"));
+    }
+
+    @Test
+    void shouldGiveValidationErrorWhenCreateDipApplicationWithoutSpecifyingAllowedExpectsFutureIncomeDecreaseReason() throws Exception {
```
Comment (KastorDriver) [inline src/test/java/com/selina/lending/api/controller/DIPControllerValidationTest.java:654]: I wonder if the name of the test should be 
shouldGiveValidationErrorWhenCreateDipApplicationWithoutSpecifying**Not**AllowedExpectsFutureIncomeDecreaseReason

**Comment by WeiLu1** [comment]
> /lgtm
> 
> Shall we have an extra test that checks the situation when the `expectsFutureIncomeDecrease = true` but the `expectsFutureIncomeDecreaseReason` is not specified at all?

Can write a test, we should expect for a validation error to be thrown in this case.

**Comment by selina-bot-user** [comment]
New changes are detected. LGTM label has been removed.

**Comment by selina-bot-user** [comment]
:star: PR built and available in a preview **selina-finance-ms-lending-api-pr-163** [here](http://ms-lending-api-pr163.selinadev.co.uk)

**Comment by selina-bot-user** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *[KastorDriver](https://github.com/Selina-Finance/ms-lending-api/pull/163#issuecomment-1506676724 "Approved")*

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=Selina-Finance%2Fms-lending-api).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/Selina-Finance/ms-lending-api/blob/main/OWNERS)~~ [KastorDriver]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by selina-bot-user** [comment]
:star: PR built and available in a preview **selina-finance-ms-lending-api-pr-163** [here](http://ms-lending-api-pr163.selinadev.co.uk)

**Comment by selina-bot-user** [comment]
:star: PR built and available in a preview **selina-finance-ms-lending-api-pr-163** [here](http://ms-lending-api-pr163.selinadev.co.uk)

**Pull request merged**